src/test/java/com/exasol/spark/s3/S3CleanupIT.java

```diff
@@ -65,52 +71,54 @@ private void assertThatBucketIsEmpty() {
         assertThat(isBucketEmpty(DEFAULT_BUCKET_NAME), equalTo(true));
     }
 
-    @Test
-    void testSourceSuccessJobEndCleanup() {
-        final MapFunction<Row, String> firstString = row -> row.getString(0);
-        final Dataset<String> df = spark.read() //
+    private Dataset<Row> getSparkDataset() {
+        return spark.read() //
                 .format("exasol-s3") //
                 .option("table", table.getFullyQualifiedName()) //
                 .options(getSparkOptions()) //
-                .load() //
-                .map(firstString, Encoders.STRING());
+                .load();
+    }
+
+    @Test
+    void testSourceSuccessJobEndCleanup() {
+        final MapFunction<Row, String> firstString = row -> row.getString(0);
+        final Dataset<String> df = getSparkDataset().map(firstString, Encoders.STRING());
         assertThat(df.collectAsList(), contains("1", "2", "3"));
         assertThatBucketIsEmpty();
     }
 
     @Test
     void testSourceSingleMapTaskFailureJobEndCleanup() {
-        final FailMax fail = new FailMax(1);
         final MapFunction<Row, Integer> failOnValue1 = row -> {
             final int value = Integer.valueOf(row.getString(0));
-            fail.on(value == 1, "The filter task value '" + value + "'.");
+            final String message = "The filter task value '" + value + "'.";
+            synchronized (TaskFailureStateCounter.class) {
```

Comment:
can you use a singleton instead, then you maybe can move the synchronized into the get_instance class method